pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,773 @@
+	if err != nil {
+		return fmt.Errorf("failed to delete VolumeSnapshotContent %s from API server: %q", content.Name, err)
+	}
+
+	return nil
+}
+
+func (ctrl *csiSnapshotController) bindandUpdateVolumeSnapshot(snapshotContent *crdv1.VolumeSnapshotContent, snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshot, error) {
+	glog.V(4).Infof("bindandUpdateVolumeSnapshot for snapshot [%s]: snapshotContent [%s]", snapshot.Name, snapshotContent.Name)
+	snapshotObj, err := ctrl.clientset.VolumesnapshotV1alpha1().VolumeSnapshots(snapshot.Namespace).Get(snapshot.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error get snapshot %s from api server: %v", snapshotKey(snapshot), err)
+	}
+
+	// Copy the snapshot object before updating it
+	snapshotCopy := snapshotObj.DeepCopy()
+
+	if snapshotObj.Spec.SnapshotContentName == snapshotContent.Name {
+		glog.Infof("bindVolumeSnapshotContentToVolumeSnapshot: VolumeSnapshot %s already bind to volumeSnapshotContent [%s]", snapshot.Name, snapshotContent.Name)
+	} else {
+		glog.Infof("bindVolumeSnapshotContentToVolumeSnapshot: before bind VolumeSnapshot %s to volumeSnapshotContent [%s]", snapshot.Name, snapshotContent.Name)
+		snapshotCopy.Spec.SnapshotContentName = snapshotContent.Name
+		updateSnapshot, err := ctrl.clientset.VolumesnapshotV1alpha1().VolumeSnapshots(snapshot.Namespace).Update(snapshotCopy)
+		if err != nil {
+			glog.Infof("bindVolumeSnapshotContentToVolumeSnapshot: Error binding VolumeSnapshot %s to volumeSnapshotContent [%s]. Error [%#v]", snapshot.Name, snapshotContent.Name, err)
+			return nil, fmt.Errorf("error updating snapshot object %s on the API server: %v", snapshotKey(updateSnapshot), err)
+		}
+		snapshotCopy = updateSnapshot
+		_, err = ctrl.storeSnapshotUpdate(snapshotCopy)
+		if err != nil {
+			glog.Errorf("%v", err)
+		}
+	}
+
+	glog.V(5).Infof("bindandUpdateVolumeSnapshot for snapshot completed [%#v]", snapshotCopy)
+	return snapshotCopy, nil
+}
+
+// UpdateSnapshotStatus converts snapshot status to crdv1.VolumeSnapshotCondition
+func (ctrl *csiSnapshotController) updateSnapshotStatus(snapshot *crdv1.VolumeSnapshot, csistatus *csi.SnapshotStatus, timestamp time.Time, bound bool) (*crdv1.VolumeSnapshot, error) {
+	glog.V(4).Infof("updating VolumeSnapshot[]%s, set status %v, timestamp %v", snapshotKey(snapshot), csistatus, timestamp)
+	status := snapshot.Status
+	change := false
+	timeAt := &metav1.Time{
+		Time: timestamp,
+	}
+
+	snapshotClone := snapshot.DeepCopy()
+	switch csistatus.Type {
+	case csi.SnapshotStatus_READY:
+		if bound {
+			status.Ready = true
+			change = true
+		}
+		if status.CreationTime == nil {
+			status.CreationTime = timeAt
+			change = true
+		}
+	case csi.SnapshotStatus_ERROR_UPLOADING:
+		if status.Error == nil {
+			status.Error = &storage.VolumeError{
+				Time:    *timeAt,
+				Message: "Failed to upload the snapshot",
```

Comment:
I think that this should be propagated also as an event.

Reply:
Sure.

Reply:
Done